File: src/webview-context-menu.ts
/**
 * Hide the webview's native right-click menu (WebView2 Back/Reload/Inspect,
 * WKWebView/WebKitGTK Reload) unless hidden debug mode turns it back on.
 *
 * A matching Tauri initialization script installs the same listener before
 * page JS so the native menu cannot flash during load.
 */

declare global {
  interface Window {
    __FREEACE_NATIVE_CONTEXT_MENU_GUARD__?: boolean;
    __FREEACE_ALLOW_NATIVE_CONTEXT_MENU__?: boolean;
  }
}

export function setNativeWebviewContextMenuAllowed(allowed: boolean): void {
  window.__FREEACE_ALLOW_NATIVE_CONTEXT_MENU__ = allowed;
}

export function isNativeWebviewContextMenuAllowed(): boolean {
  return window.__FREEACE_ALLOW_NATIVE_CONTEXT_MENU__ === true;
}

export function installNativeWebviewContextMenuGuard(): void {
  if (window.__FREEACE_NATIVE_CONTEXT_MENU_GUARD__) return;
  window.__FREEACE_NATIVE_CONTEXT_MENU_GUARD__ = true;
  document.addEventListener(
    "contextmenu",
    (event) => {
      if (!window.__FREEACE_ALLOW_NATIVE_CONTEXT_MENU__) {
        event.preventDefault();
      }
    },
    true,
  );
}

File: src-tauri/src/launch/webview_context_menu.rs
//! Early-load script that blocks the native webview context menu until JS
//! opts in via `window.__FREEACE_ALLOW_NATIVE_CONTEXT_MENU__` (hidden debug mode).

/// Same listener the frontend installs; runs before page scripts so WebView2
/// / WKWebView / WebKitGTK cannot flash Back, Reload, or Inspect on first right-click.
pub(crate) const NATIVE_CONTEXT_MENU_GUARD_SCRIPT: &str = r#"
(function () {
  if (window.__FREEACE_NATIVE_CONTEXT_MENU_GUARD__) return;
  window.__FREEACE_NATIVE_CONTEXT_MENU_GUARD__ = true;
  document.addEventListener("contextmenu", function (event) {
    if (!window.__FREEACE_ALLOW_NATIVE_CONTEXT_MENU__) {
      event.preventDefault();
    }
  }, true);
})();
"#;
